reactfire/useObservable/index.ts

```diff
@@ -1,89 +1,46 @@
-import { first, startWith } from 'rxjs/operators';
-import { ActiveRequest, ObservablePromiseCache } from './requestCache';
+import { BehaviorReplaySubject } from './behaviorReplaySubject';
 
-const requestCache = new ObservablePromiseCache();
-
-export function preloadRequest(
-  getPromise,
-  requestId: string
-): { requestId: string; request: ActiveRequest } {
-  const request = requestCache.createDedupedRequest(getPromise, requestId);
-
-  return {
-    requestId: requestId,
-    request
-  };
-}
+const DEFAULT_TIMEOUT = 30_000;
+const preloadedObservables = new Map<string, BehaviorReplaySubject<unknown>>();
 
 // Starts listening to an Observable.
 // Call this once you know you're going to render a
 // child that will consume the observable
-export function preloadObservable(
-  observable$: Observable<any>,
-  observableId: string
-): { requestId: string; request: ActiveRequest } {
-  return preloadRequest(
-    () => observable$.pipe(first()).toPromise(),
-    observableId
-  );
-}
-
-export function usePreloadedRequest(preloadResult: { requestId: string }) {
-  const request = requestCache.getRequest(preloadResult.requestId);
-
-  // Suspend if we're not ready yet
-  if (!request.isComplete) {
-    throw request.promise;
-  }
-
-  if (request.error) {
-    throw request.error;
+export function preloadObservable<T>(source: Observable<T>, id: string) {
+  if (preloadedObservables.has(id)) {
+    return preloadedObservables.get(id) as BehaviorReplaySubject<T>;
+  } else {
+    const observable = new BehaviorReplaySubject(source, DEFAULT_TIMEOUT);
+    preloadedObservables.set(id, observable);
+    return observable;
   }
-
-  return request.value;
 }
 
 export function useObservable<T>(
-  observable$: Observable<T | any>,
+  source: Observable<T | any>,
   observableId: string,
   startWithValue?: T | any
 ): T {
   if (!observableId) {
     throw new Error('cannot call useObservable without an observableId');
   }
-
-  const result = preloadObservable(observable$, observableId);
-
-  let initialValue = startWithValue;
-  if (initialValue === undefined) {
-    // this will Suspend until the Promise resolves
-    initialValue = usePreloadedRequest(result);
+  const observable = preloadObservable(source, observableId);
+  if (!observable.hasValue && !startWithValue) {
+    throw observable.firstEmission;
   }
-
-  const [latestValue, setValue] = React.useState(initialValue);
-
+  const [latest, setValue] = React.useState(
+    observable.hasValue ? observable.value : startWithValue
+  );
   React.useEffect(() => {
-    const subscription = observable$.pipe(startWith(initialValue)).subscribe(
-      newVal => {
-        // update the value in requestCache
-        result.request.setValue(newVal);
-
-        // update state
-        setValue(newVal);
-      },
-      error => {
-        console.error('There was an error', error);
-        throw error;
+    const subscription = observable.subscribe(
+      v => setValue(v),
```

Comment:
Same as above, if you pass a function to setValue, [React will apparently also try to call it](https://reactjs.org/docs/hooks-reference.html#functional-updates)

So I think we need to do:

```ts
v => setValue(() => v)
```